util/log/log.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package log provides log utils.
+package log
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AddObjectHierarchy adds the hierarchy of an object as k/v pairs to the logger in ctx.
+func AddObjectHierarchy(ctx context.Context, c client.Client, obj metav1.Object) (context.Context, logr.Logger, error) {
```

Comment:
Would be good to get some feedback about that func first.

I would then:
* add corresponding unit tests
* document it in the v1.2=>v1.3 doc that providers can use it too

I intentionally:
* didn't resolve the entire ownerRef tree (not sure how deep they are in folks using CAPI in non-mainstream ways)
* did resolve MachineSets as that's a known case and very useful to include MachineDeployments in Machine logs
* did use the typed client to get the MachineSet as the MachineSet will be automatically cached, it would have been a lot more complicated to require an unstructured cached client
* didn't include an allow-list of ownerRefs we want to add as k/v pairs. We could do that but I'm not sure if it's okay to hard-code "KubeadmControlPlane" (things like "MachineDeployment", "MachineSet" should be fine.

This change gives us the following k/v pairs:

```md
* bootstrap
  * KubeadmConfig
    * kv: Cluster | MachinePool | KubeadmConfig
    * kv: Cluster | KubeadmControlPlane | Machine | KubeadmConfig
    * kv: Cluster | MachineDeployment | MachineSet | Machine | KubeadmConfig
* controlplane
  * KubeadmControlPlane
    * kv: Cluster | KubeadmControlPlane
* core
  * Cluster
    * kv: Cluster 
  * MachineDeployment
    * kv: Cluster | MachineDeployment
  * MachineSet
    * kv: Cluster | MachineDeployment | MachineSet
  * Machine
    * kv: Cluster | MachineDeployment | MachineSet | Machine
    * kv: Cluster | KubeadmControlPlane | Machine
  * MachineHealthcheck
    * kv: Cluster | MachineHealthcheck

* CAPD
  * DockerCluster
    * kv: Cluster | DockerCluster
  * DockerMachine
    * kv: Cluster | MachineDeployment | MachineSet | Machine | DockerMachine
    * kv: Cluster | KubeadmControlPlane | Machine | DockerMachine 
  * DockerMachinePool
    * kv: Cluster | MachinePool | DockerMachinePool
```

Reply:
cc @fabriziopandini

Reply:
The result is what we expect, so I'm +1 for the change with the current implementation (we can eventually improve later)
The only improvement that comes to my mind is to avoid to add twice the same owner ref, that is something that could happen e.g when the topology controllers adds additional owner refs to ensure object cleanup

Reply:
This is safe from the logger-side. The log k/v pairs are deduplicated by `key`

Reply:
I'll add it here anyway. Probably better to deduplicate here, vs otherwise it will be done on every log call